packages/Search/src/details/DTK_DetailsUtils.hpp

```diff
@@ -18,62 +18,92 @@
 namespace DataTransferKit
 {
 
-/** \brief Assigns the given value to all elements in the view.
- */
 template <typename T, typename DeviceType>
-void fill( Kokkos::View<T *, DeviceType> out, T const &value )
+class ExclusiveScanFunctor
 {
-    using ExecutionSpace = typename DeviceType::execution_space;
-    Kokkos::parallel_for(
-        "fill", Kokkos::RangePolicy<ExecutionSpace>( 0, out.extent( 0 ) ),
-        KOKKOS_LAMBDA( int i ) { out( i ) = value; } );
-    Kokkos::fence();
-}
+  public:
+    ExclusiveScanFunctor( Kokkos::View<T *, DeviceType> const &in,
+                          Kokkos::View<T *, DeviceType> const &out )
+        : _in( in )
+        , _out( out )
+    {
+    }
+    KOKKOS_INLINE_FUNCTION void operator()( int i, T &update,
+                                            bool final_pass ) const
+    {
+        T const in_i = _in( i );
+        if ( final_pass )
+            _out( i ) = update;
+        update += in_i;
+    }
+
+  private:
+    Kokkos::View<T *, DeviceType> _in;
+    Kokkos::View<T *, DeviceType> _out;
+};
 
 /** \brief Computes an exclusive scan.
  *
- *  When \c out is not provided or if \c in and \c out are the same view, the
+ *  When \c dst is not provided or if \c src and \c dst are the same view, the
  *  scan is performed in-place.
  *
- *  \pre \c in and \c out must have the same size.
+ *  \pre \c src and \c dst must be of rank 1 and have the same size.
  */
-template <typename T, typename DeviceType>
-void exclusivePrefixSum(
-    Kokkos::View<T *, DeviceType> in,
-    Kokkos::View<T *, DeviceType> out = Kokkos::View<T *, DeviceType>() )
+template <typename ST, typename... SP, typename DT, typename... DP>
+void exclusivePrefixSum( Kokkos::View<ST, SP...> const &src,
+                         Kokkos::View<DT, DP...> const &dst )
```

Comment:
`dst` is also a `cons ref`? Looks like the function has no output.

Reply:
Yes, view is const but values are not.  `Kokkos::deep_copy()` also takes both views as const ref.

Reply:
I know, still ugly to read. It looks like there is no output.

Reply:
I added a parameter description with `[in]` and `[out]` specifiers.